From 63f4c5ba9d46604da4d4d6e81ea82aff27e3ff2c Mon Sep 17 00:00:00 2001
From: Jian-Ming Wang <jian.ming.wang@mitaccomputing.com>
Date: Wed, 14 May 2025 03:39:41 +0000
Subject: [PATCH 3/3] Rewrite early_vga based on fbida 2.14 and for the change
 of data type to xyz.openbmc_project.State.Boot.Raw.

Upstream-Status: Inappropriate [No upstream maintenance]
---
 early_vga.c | 842 ++++++++++++++++++++++++++++++----------------------
 1 file changed, 487 insertions(+), 355 deletions(-)

diff --git a/early_vga.c b/early_vga.c
index 8fd030b..180fe64 100644
--- a/early_vga.c
+++ b/early_vga.c
@@ -43,9 +43,20 @@
 #include "desktop.h"
 #include "fbiconfig.h"
 
-#include "transupp.h"		/* Support routines for jpegtran */
+#include "transupp.h"           /* Support routines for jpegtran */
 #include "jpegtools.h"
 
+#include <dbus/dbus.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <netinet/in.h>
+#include <netinet/ether.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+
+#define POSTCODE_INFO_LEN 20
+#define BMC_ETH_INFO_LEN 50
+
 /* ---------------------------------------------------------------------- */
 
 /* variables for read_image */
@@ -128,9 +139,9 @@ static void
 version(void)
 {
     fprintf(stdout,
-	    "fbi version " VERSION ", compiled on %s\n"
-	    "(c) 1998-2016 Gerd Hoffmann <gerd@kraxel.org>\n",
-	    __DATE__ );
+            "fbi version " VERSION ", compiled on %s\n"
+            "(c) 1998-2016 Gerd Hoffmann <gerd@kraxel.org>\n",
+            __DATE__ );
 }
 
 static void
@@ -139,28 +150,28 @@ usage(FILE *fp, char *name)
     char           *h;
 
     if (NULL != (h = strrchr(name, '/')))
-	name = h+1;
+        name = h+1;
     fprintf(fp,
-	    "\n"
-	    "This program displays images using the Linux framebuffer device.\n"
-	    "Supported formats: PhotoCD, jpeg, ppm, gif, tiff, xwd, bmp, png,\n"
-	    "webp. It tries to use ImageMagick's convert for unknown file formats.\n"
-	    "\n"
-	    "usage: %s [ options ] file1 file2 ... fileN\n"
-	    "\n",
-	    name);
+            "\n"
+            "This program displays images using the Linux framebuffer device.\n"
+            "Supported formats: PhotoCD, jpeg, ppm, gif, tiff, xwd, bmp, png,\n"
+            "webp. It tries to use ImageMagick's convert for unknown file formats.\n"
+            "\n"
+            "usage: %s [ options ] file1 file2 ... fileN\n"
+            "\n",
+            name);
 
     cfg_help_cmdline(fp,fbi_cmd,4,20,0);
     cfg_help_cmdline(fp,fbi_cfg,4,20,40);
 
     fprintf(fp,
-	    "\n"
-	    "Large images can be scrolled using the cursor keys.  Zoom in/out\n"
-	    "works with '+' and '-'.  Use ESC or 'q' to quit.  Space and PgDn\n"
-	    "show the next, PgUp shows the previous image. Jumping to a image\n"
-	    "works with <i>g.  Return acts like Space but additionally prints\n"
-	    "prints the filename of the currently displayed image to stdout.\n"
-	    "\n");
+            "\n"
+            "Large images can be scrolled using the cursor keys.  Zoom in/out\n"
+            "works with '+' and '-'.  Use ESC or 'q' to quit.  Space and PgDn\n"
+            "show the next, PgUp shows the previous image. Jumping to a image\n"
+            "works with <i>g.  Return acts like Space but additionally prints\n"
+            "prints the filename of the currently displayed image to stdout.\n"
+            "\n");
 }
 
 /* ---------------------------------------------------------------------- */
@@ -183,7 +194,7 @@ static struct flist *flist_find(const char *filename)
     struct flist *f;
 
     list_for_each(item,&flist) {
-	f = list_entry(item, struct flist, list);
+        f = list_entry(item, struct flist, list);
         if (strcmp(filename, f->name) == 0)
             return f;
     }
@@ -198,14 +209,14 @@ static int flist_add_list(const char *listfile)
 
     list = fopen(listfile,"r");
     if (NULL == list) {
-	fprintf(stderr,"open %s: %s\n",listfile,strerror(errno));
-	return -1;
+        fprintf(stderr,"open %s: %s\n",listfile,strerror(errno));
+        return -1;
     }
     fstat(fileno(list), &liststat);
     while (NULL != fgets(filename,sizeof(filename)-1,list)) {
-	size_t off = strcspn(filename,"\r\n");
-	if (off)
-	    filename[off] = 0;
+        size_t off = strcspn(filename,"\r\n");
+        if (off)
+            filename[off] = 0;
         f = flist_find(filename);
         if (!f) {
             f = flist_add(filename);
@@ -231,8 +242,8 @@ static void flist_renumber(void)
     int i = 0;
 
     list_for_each(item,&flist) {
-	f = list_entry(item, struct flist, list);
-	f->nr = ++i;
+        f = list_entry(item, struct flist, list);
+        f->nr = ++i;
     }
     fcount = i;
 }
@@ -256,7 +267,7 @@ static int flist_check_reload_list(const char *listfile)
         return ret;
 
     list_for_each_safe(item, safe, &flist) {
-	f = list_entry(item, struct flist, list);
+        f = list_entry(item, struct flist, list);
         if (f->list_check) {
             f->list_check = false;
         } else {
@@ -273,52 +284,46 @@ static int flist_islast(struct flist *f)
     return (&flist == f->list.next) ? 1 : 0;
 }
 
+/*
 static int flist_isfirst(struct flist *f)
 {
     return (&flist == f->list.prev) ? 1 : 0;
 }
+*/
 
 static struct flist* flist_first(void)
 {
     return list_entry(flist.next, struct flist, list);
 }
 
+/*
 static struct flist* flist_last(void)
 {
     return list_entry(flist.prev, struct flist, list);
 }
+*/
 
 static struct flist* flist_next(struct flist *f, int eof, int loop)
 {
     if (flist_islast(f)) {
-	if (eof)
-	    return NULL;
-	if (loop)
-	    return flist_first();
-	return f;
+        if (eof)
+            return NULL;
+        if (loop)
+            return flist_first();
+        return f;
     }
     return list_entry(f->list.next, struct flist, list);
 }
 
-static struct flist* flist_prev(struct flist *f, int loop)
-{
-    if (flist_isfirst(f)) {
-	if (loop)
-	    return flist_last();
-	return f;
-    }
-    return list_entry(f->list.prev, struct flist, list);
-}
-
 static struct flist* flist_goto(int dest)
 {
     struct list_head *item;
     struct flist *f;
 
     list_for_each(item,&flist) {
-	f = list_entry(item, struct flist, list);
-	if (f->nr == dest)
-	    return f;
+        f = list_entry(item, struct flist, list);
+        if (f->nr == dest)
+            return f;
     }
     return NULL;
 }
@@ -331,10 +336,10 @@ static void flist_randomize(void)
     srand((unsigned)time(NULL));
     flist_renumber();
     for (count = fcount; count > 0; count--) {
-	f = flist_goto((rand() % count)+1);
-	list_del(&f->list);
-	list_add_tail(&f->list,&flist);
-	flist_renumber();
+        f = flist_goto((rand() % count)+1);
+        list_del(&f->list);
+        list_add_tail(&f->list,&flist);
+        flist_renumber();
     }
 }
 
@@ -344,9 +349,9 @@ static void flist_print_tagged(FILE *fp)
     struct flist *f;
 
     list_for_each(item,&flist) {
-	f = list_entry(item, struct flist, list);
-	if (f->tag)
-	    fprintf(fp,"%s\n",f->name);
+        f = list_entry(item, struct flist, list);
+        if (f->tag)
+            fprintf(fp,"%s\n",f->name);
     }
 }
 
@@ -354,21 +359,21 @@ static void flist_print_tagged(FILE *fp)
 
 static void
 shadow_draw_image(struct ida_image *img, int xoff, int yoff,
-		  unsigned int first, unsigned int last, int weight)
+                  unsigned int first, unsigned int last, int weight)
 {
     unsigned int     xs, ys;
 
     if (100 == weight)
-	shadow_clear_lines(first, last);
+        shadow_clear_lines(first, last);
     else
-	shadow_darkify(0, gfx->hdisplay-1, first, last, 100 - weight);
+        shadow_darkify(0, gfx->hdisplay-1, first, last, 100 - weight);
 
     /* image < screen: center image */
     xs = 0, ys = 0;
     if (img->i.width < gfx->hdisplay)
-	xs += (gfx->hdisplay - img->i.width) / 2;
+        xs += (gfx->hdisplay - img->i.width) / 2;
     if (img->i.height < gfx->vdisplay)
-	ys += (gfx->vdisplay - img->i.height) / 2;
+        ys += (gfx->vdisplay - img->i.height) / 2;
 
     shadow_composite_image(img, xs - xoff, ys - yoff, weight);
 }
@@ -380,12 +385,12 @@ static void status_prepare(void)
     int y2 = gfx->vdisplay - 1;
 
     if (img) {
-	shadow_draw_image(img, fcurrent->left, fcurrent->top, y1, y2, 100);
-	shadow_darkify(0, gfx->hdisplay-1, y1, y2, transparency);
+        shadow_draw_image(img, fcurrent->left, fcurrent->top, y1, y2, 100);
+        shadow_darkify(0, gfx->hdisplay-1, y1, y2, transparency);
     } else {
-	shadow_clear_lines(y1, y2);
+        shadow_clear_lines(y1, y2);
     }
-    shadow_draw_line(0, gfx->hdisplay-1, y1-1, y1-1);
+    //shadow_draw_line(0, gfx->hdisplay-1, y1-1, y1-1);
 }
 
 static void status_update(unsigned char *desc, char *info)
@@ -394,15 +399,15 @@ static void status_update(unsigned char *desc, char *info)
     char str[128];
 
     if (!statusline)
-	return;
+        return;
     status_prepare();
-
+    
     snprintf(str,ARRAY_SIZE(str),"%s",desc);
     shadow_draw_string(0, yt, str, -1);
     if (info) {
-	snprintf(str,ARRAY_SIZE(str), "[ %s ] H - Help", info);
+        snprintf(str,ARRAY_SIZE(str), "%s", info);
     } else {
-	snprintf(str,ARRAY_SIZE(str), "| H - Help");
+        snprintf(str,ARRAY_SIZE(str), "");
     }
     shadow_draw_string(gfx->hdisplay, yt, str, 1);
 
@@ -423,50 +428,202 @@ static void status_error(unsigned char *msg)
     sleep(2);
 }
 
-static void show_help(void)
+static char* get_network_intf_info()
 {
-    static char *help[] = {
-	"keyboard commands",
-	"~~~~~~~~~~~~~~~~~",
-	"  cursor keys    - scroll image",
-	"  PgUp, k        - previous image",
-	"  PgDn, SPACE, j - next image",
-	"  <i>g           - jump to image #i",
-	"",
-	"  a              - autozoom image",
-	"  +/-            - zoom in/out",
-	"  <i>s           - set zoom to <i>%",
-	"",
-	"  ESC, q         - quit",
-	"  v              - toggle statusline",
-	"  h              - show this help text",
-	"  i              - show EXIF info",
-	"  p              - pause slideshow",
-	"",
-	"available if started with --edit switch,",
-	"rotation works for jpeg images only:",
-	"  D, Shift+d     - delete image",
-	"  r              - rotate 90 degrees clockwise",
-	"  l              - rotate 90 degrees counter-clockwise",
-	"  x              - mirror image vertically (top / bottom)",
-	"  y              - mirror image horizontally (left to right)",
-    };
-
-    shadow_draw_text_box(24, 16, transparency,
-			 help, ARRAY_SIZE(help));
-    shadow_render(gfx);
+    int fd;
+    boolean has_mac = 0;
+    unsigned char mac_address[6];
+    struct ifreq ifr;
+    char* str = malloc(BMC_ETH_INFO_LEN);
+
+    if (str == NULL)
+        return NULL;
+
+    fd = socket(AF_INET, SOCK_DGRAM, 0);
+
+    ifr.ifr_addr.sa_family = AF_INET;
+    strncpy(ifr.ifr_name, "bond0", IFNAMSIZ-1);
+    if (ioctl(fd, SIOCGIFHWADDR, &ifr) == 0) {
+        memcpy(mac_address, ifr.ifr_hwaddr.sa_data, 6);
+        has_mac = 1;
+    }
+    if (ioctl(fd, SIOCGIFADDR, &ifr) == 0) {
+        if (has_mac) {
+                snprintf(str, BMC_ETH_INFO_LEN, "BMC IP: %s, MAC: %s", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr), ether_ntoa((struct ether_addr *) mac_address));
+        } else {
+                snprintf(str, BMC_ETH_INFO_LEN, "BMC IP: %s", inet_ntoa(((struct sockaddr_in *)&ifr.ifr_addr)->sin_addr));
+        }
+    } else {
+    	snprintf(str, BMC_ETH_INFO_LEN, "BMC Network Unplugged");
+    }
+    close(fd);
+    return (char*) str;
 }
 
-/* ---------------------------------------------------------------------- */
+
+static uint64_t update_post_code(unsigned char *desc, char *info)
+{
+    char postcode_buffer[POSTCODE_INFO_LEN];
+    char bmc_eth_info_buffer[BMC_ETH_INFO_LEN];
+    DBusError err;
+    DBusConnection* conn;
+    DBusMessage* msg;
+    DBusMessageIter args;
+    int ret;
+
+    char* bmc_eth_info = get_network_intf_info();
+    if (bmc_eth_info != NULL) {
+        snprintf(bmc_eth_info_buffer, sizeof(bmc_eth_info_buffer), "%s", bmc_eth_info);
+        free(bmc_eth_info);
+    }
+    status_update("Dbus Init", bmc_eth_info_buffer);
+    dbus_error_init(&err);
+    // connect to the bus
+    conn = dbus_bus_get(DBUS_BUS_SYSTEM, &err);
+    if (dbus_error_is_set(&err)) { 
+        fprintf(stderr, "Connection Error (%s)\n", err.message); 
+        dbus_error_free(&err); 
+        printf("Failed at dbus_bus_get.\n"); 
+        status_update(err.message, bmc_eth_info_buffer);
+        shadow_render(gfx);
+    }
+    if (NULL == conn) {
+        printf("Got NULL DBusConnection.\n"); 
+        exit(1); 
+    }
+    ret = dbus_bus_request_name(conn, "test.method.server",
+            DBUS_NAME_FLAG_REPLACE_EXISTING,
+            &err);
+    if (dbus_error_is_set(&err)) {
+        fprintf(stderr, "Name Error (%s)\n", err.message);
+        dbus_error_free(&err);
+        printf("Failed at dbus_bus_request_name.\n"); 
+    }
+    if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != ret) {
+        printf("Error: Not DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER.\n"); 
+        exit(1);
+    }
+    // add a rule for which messages we want to see
+#define rule_post_code "type='signal',path='/xyz/openbmc_project/state/boot/raw0',interface='org.freedesktop.DBus.Properties'"
+#define rule_restart_cause "type='signal',path='/xyz/openbmc_project/state/host0',interface='org.freedesktop.DBus.Properties'"
+    dbus_bus_add_match(conn, 
+         rule_post_code, 
+         &err); // see signals from the given interface
+    dbus_bus_add_match(conn, 
+         rule_restart_cause, 
+         &err); // see signals from the given interface
+    dbus_connection_flush(conn);
+    if (dbus_error_is_set(&err)) { 
+        fprintf(stderr, "Match Error (%s)\n", err.message);
+        exit(1); 
+    }
+
+    snprintf(postcode_buffer, sizeof(postcode_buffer), "Post Code: Waiting BIOS");
+    status_update(postcode_buffer, bmc_eth_info_buffer);
+    shadow_render(gfx);
+    while (1){
+        dbus_connection_read_write(conn, 0);
+        msg = dbus_connection_pop_message(conn);
+        if (msg == NULL) {
+            sleep(1);
+            continue;
+        }
+        if (dbus_message_is_signal(msg, "org.freedesktop.DBus.Properties", "PropertiesChanged")) {
+            if (!dbus_message_iter_init(msg, &args))
+                fprintf(stderr, "Message has no arguments!\n");
+
+            /*
+             *  Arg 1: String
+             *  Arg 2: Array of DBUS_TYPE_DICT_ENTRY
+             *      Dict: <DBUS_TYPE_STRING:DBUS_TYPE_VARIANT>
+             *                              DBUS_TYPE_STRUCT
+             *  Arg 3: Array		
+             *
+             */
+            while ((dbus_message_iter_get_arg_type (&args)) != DBUS_TYPE_INVALID) {
+                int argType = dbus_message_iter_get_arg_type (&args);
+                if (argType == DBUS_TYPE_ARRAY) {
+                    int elementType = dbus_message_iter_get_element_type(&args);
+                    if (elementType == DBUS_TYPE_DICT_ENTRY) {
+                        DBusMessageIter dict;
+                        dbus_message_iter_recurse(&args, &dict);
+                        char *key;
+                        while (dbus_message_iter_get_arg_type (&dict) == DBUS_TYPE_DICT_ENTRY) {
+                            DBusMessageIter entry;
+                            dbus_message_iter_recurse(&dict, &entry);
+                            dbus_message_iter_get_basic(&entry, &key);
+                            dbus_message_iter_next(&entry);
+                            // key represent the property name. It is better to check interface name as well.
+                            if (!strcmp(key, "Value")){
+                                if (dbus_message_iter_get_arg_type (&entry) == DBUS_TYPE_VARIANT) {
+                                    DBusMessageIter var;
+                                    dbus_message_iter_recurse(&entry, &var);
+                                    if (dbus_message_iter_get_arg_type(&var) == DBUS_TYPE_STRUCT) {
+                                        DBusMessageIter structIter;
+                                        dbus_message_iter_recurse(&var, &structIter);
+
+                                        const int post_code_container_type = dbus_message_iter_get_arg_type(&structIter);
+                                        if ( post_code_container_type == DBUS_TYPE_UINT64 ){
+                                            // Legacy
+                                            uint64_t post_code;
+                                            dbus_message_iter_get_basic(&structIter, &post_code);
+                                            snprintf(postcode_buffer, sizeof(postcode_buffer), "%s: 0x%04X", "Post Code", post_code);
+                                            status_update(postcode_buffer, bmc_eth_info_buffer);
+                                            shadow_render(gfx);
+                                        } else if ( post_code_container_type == DBUS_TYPE_ARRAY ){
+                                            // New
+                                            DBusMessageIter array;
+                                            dbus_message_iter_recurse(&structIter, &array);
+
+                                            const int post_code_element_type = dbus_message_iter_get_arg_type(&array);
+                                            if ( post_code_element_type == DBUS_TYPE_BYTE ){
+                                                uint8_t *intArrayArg;
+                                                int elementNum;
+                                                dbus_message_iter_get_fixed_array(&array, &intArrayArg, &elementNum);
+                                                snprintf(postcode_buffer, sizeof(postcode_buffer), "%s: 0x%02X%02X", "Post Code", *(intArrayArg+1), *intArrayArg);
+                                                status_update(postcode_buffer, bmc_eth_info_buffer);
+                                                shadow_render(gfx);
+                                            } else {
+                                                fprintf(stderr, "%s:%d Invalid Data type for element of post code\n", __func__, __LINE__);
+                                            }
+                                        } else {
+                                            fprintf(stderr, "%s:%d Invalid Data type for post code\n", __func__, __LINE__);
+                                        }
+                                    }
+                                } else {
+                                    fprintf(stderr, "%s:%d Checkpoint\n", __func__, __LINE__);
+                                }
+                            } else if (!strcmp(key, "CurrentHostState")){
+                                const char* bmc_eth_info = get_network_intf_info();
+                                if (bmc_eth_info != NULL) {
+                                    snprintf(bmc_eth_info_buffer, sizeof(bmc_eth_info_buffer), "%s", bmc_eth_info);
+                                    free(bmc_eth_info);
+                                }
+                                snprintf(postcode_buffer, sizeof(postcode_buffer), "Post Code: Waiting BIOS");
+                                status_update(postcode_buffer, bmc_eth_info_buffer);
+                                shadow_render(gfx);
+                            } else {
+                                fprintf(stderr, "%s:%d Key=%s\n", __func__, __LINE__, key);
+                            }
+                            dbus_message_iter_next (&dict);
+                        }
+                    }
+                }
+                dbus_message_iter_next (&args);
+            }
+        }
+        dbus_message_unref(msg);	    
+    }
+}
 
 static void free_image(struct ida_image *img)
 {
     if (img) {
-	if (img->p) {
-	    img_mem -= img->i.width * img->i.height * 3;
-	    ida_image_free(img);
-	}
-	free(img);
+        if (img->p) {
+            img_mem -= img->i.width * img->i.height * 3;
+            ida_image_free(img);
+        }
+        free(img);
     }
 }
 
@@ -483,8 +640,8 @@ read_image(char *filename)
 
     /* open file */
     if (NULL == (fp = fopen(filename, "r"))) {
-	fprintf(stderr,"open %s: %s\n",filename,strerror(errno));
-	return NULL;
+        fprintf(stderr,"open %s: %s\n",filename,strerror(errno));
+        return NULL;
     }
     memset(blk,0,sizeof(blk));
     fread(blk,1,sizeof(blk),fp);
@@ -493,37 +650,37 @@ read_image(char *filename)
     /* pick loader */
     list_for_each(item,&loaders) {
         loader = list_entry(item, struct ida_loader, list);
-	if (NULL == loader->magic)
-	    break;
-	if (0 == memcmp(blk+loader->moff,loader->magic,loader->mlen))
-	    break;
-	loader = NULL;
+        if (NULL == loader->magic)
+            break;
+        if (0 == memcmp(blk+loader->moff,loader->magic,loader->mlen))
+            break;
+        loader = NULL;
     }
     if (NULL == loader) {
-	/* no loader found, try to use ImageMagick's convert */
-	int p[2];
-
-	if (0 != pipe(p))
-	    return NULL;
-	switch (fork()) {
-	case -1: /* error */
-	    perror("fork");
-	    close(p[0]);
-	    close(p[1]);
-	    return NULL;
-	case 0: /* child */
-	    dup2(p[1], 1 /* stdout */);
-	    close(p[0]);
-	    close(p[1]);
-	    execlp("convert", "convert", "-depth", "8", filename, "ppm:-", NULL);
-	    exit(1);
-	default: /* parent */
-	    close(p[1]);
-	    fp = fdopen(p[0], "r");
-	    if (NULL == fp)
-		return NULL;
-	    loader = &ppm_loader;
-	}
+        /* no loader found, try to use ImageMagick's convert */
+        int p[2];
+
+        if (0 != pipe(p))
+            return NULL;
+        switch (fork()) {
+        case -1: /* error */
+            perror("fork");
+            close(p[0]);
+            close(p[1]);
+            return NULL;
+        case 0: /* child */
+            dup2(p[1], 1 /* stdout */);
+            close(p[0]);
+            close(p[1]);
+            execlp("convert", "convert", "-depth", "8", filename, "ppm:-", NULL);
+            exit(1);
+        default: /* parent */
+            close(p[1]);
+            fp = fdopen(p[0], "r");
+            if (NULL == fp)
+                return NULL;
+            loader = &ppm_loader;
+        }
     }
 
     /* load image */
@@ -531,15 +688,15 @@ read_image(char *filename)
     memset(img,0,sizeof(*img));
     data = loader->init(fp,filename,0,&img->i,0);
     if (NULL == data) {
-	fprintf(stderr,"loading %s [%s] FAILED\n",filename,loader->name);
-	free_image(img);
-	return NULL;
+        fprintf(stderr,"loading %s [%s] FAILED\n",filename,loader->name);
+        free_image(img);
+        return NULL;
     }
     ida_image_alloc(img);
     img_mem += img->i.width * img->i.height * 3;
     for (y = 0; y < img->i.height; y++) {
         check_console_switch();
-	loader->read(ida_image_scanline(img, y), y, data);
+        loader->read(ida_image_scanline(img, y), y, data);
     }
     loader->done(data);
     return img;
@@ -563,16 +720,16 @@ scale_image(struct ida_image *src, float scale)
     p.height = src->i.height * scale;
     p.dpi    = src->i.dpi;
     if (0 == p.width)
-	p.width = 1;
+        p.width = 1;
     if (0 == p.height)
-	p.height = 1;
+        p.height = 1;
 
     data = desc_resize.init(src,&rect,&dest->i,&p);
     ida_image_alloc(dest);
     img_mem += dest->i.width * dest->i.height * 3;
     for (y = 0; y < dest->i.height; y++) {
         check_console_switch();
-	desc_resize.work(src, &rect, ida_image_scanline(dest, y), y, data);
+        desc_resize.work(src, &rect, ida_image_scanline(dest, y), y, data);
     }
     desc_resize.done(data);
     return dest;
@@ -584,7 +741,7 @@ static float auto_scale(struct ida_image *img)
 
     xs = (float)gfx->hdisplay / img->i.width;
     if (fitwidth)
-	return xs;
+        return xs;
     ys = (float)gfx->vdisplay / img->i.height;
     scale = (xs < ys) ? xs : ys;
     return scale;
@@ -609,141 +766,117 @@ static void effect_blend(struct flist *f, struct flist *t)
 
     gettimeofday(&start, NULL);
     do {
-	gettimeofday(&now, NULL);
-	msecs  = (now.tv_sec  - start.tv_sec)  * 1000;
-	msecs += (now.tv_usec - start.tv_usec) / 1000;
-	weight = msecs * 100 / blend_msecs;
-	if (weight > 100)
-	    weight = 100;
-	shadow_draw_image(flist_img_get(f), f->left, f->top,
-			  0, gfx->vdisplay-1, 100);
-	shadow_draw_image(flist_img_get(t), t->left, t->top,
-			  0, gfx->vdisplay-1, weight);
-
-	if (perfmon) {
-	    pos += snprintf(linebuffer+pos, sizeof(linebuffer)-pos,
-			    " %d%%", weight);
-	    status_update(linebuffer, NULL);
-	    count++;
-	}
-
-	shadow_render(gfx);
+        gettimeofday(&now, NULL);
+        msecs  = (now.tv_sec  - start.tv_sec)  * 1000;
+        msecs += (now.tv_usec - start.tv_usec) / 1000;
+        weight = msecs * 100 / blend_msecs;
+        if (weight > 100)
+            weight = 100;
+        shadow_draw_image(flist_img_get(f), f->left, f->top,
+                          0, gfx->vdisplay-1, 100);
+        shadow_draw_image(flist_img_get(t), t->left, t->top,
+                          0, gfx->vdisplay-1, weight);
+
+        if (perfmon) {
+            pos += snprintf(linebuffer+pos, sizeof(linebuffer)-pos,
+                            " %d%%", weight);
+            status_update(linebuffer, NULL);
+            count++;
+        }
+
+        shadow_render(gfx);
     } while (weight < 100);
 
     if (perfmon) {
-	gettimeofday(&now, NULL);
-	msecs  = (now.tv_sec  - start.tv_sec)  * 1000;
-	msecs += (now.tv_usec - start.tv_usec) / 1000;
-	pos += snprintf(linebuffer+pos, sizeof(linebuffer)-pos,
-			" | %d/%d -> %d msec",
-			msecs, count, msecs/count);
-	status_update(linebuffer, NULL);
-	shadow_render(gfx);
-	sleep(2);
+        gettimeofday(&now, NULL);
+        msecs  = (now.tv_sec  - start.tv_sec)  * 1000;
+        msecs += (now.tv_usec - start.tv_usec) / 1000;
+        pos += snprintf(linebuffer+pos, sizeof(linebuffer)-pos,
+                        " | %d/%d -> %d msec",
+                        msecs, count, msecs/count);
+        status_update(linebuffer, NULL);
+        shadow_render(gfx);
+        sleep(2);
     }
 }
 
 static int
 svga_show(struct flist *f, struct flist *prev,
-	  int timeout, char *desc, char *info, int *nr)
+          int timeout, char *desc, char *info, int *nr)
 {
     static int        paused = 0, skip = -1;
 
     struct ida_image  *img = flist_img_get(f);
-    int               exif = 0, help = 0;
     int               rc;
-    char              key[16];
-    uint32_t          keycode, keymod;
-    char              linebuffer[80];
 
     *nr = 0;
     if (NULL == img)
-	return skip;
+        return skip;
 
     redraw = 1;
     for (;;) {
-	if (redraw) {
-	    redraw = 0;
-	    if (img->i.height <= gfx->vdisplay) {
-		f->top = 0;
-	    } else {
-		if (f->top < 0)
-		    f->top = 0;
-		if (f->top + gfx->vdisplay > img->i.height)
-		    f->top = img->i.height - gfx->vdisplay;
-	    }
-	    if (img->i.width <= gfx->hdisplay) {
-		f->left = 0;
-	    } else {
-		if (f->left < 0)
-		    f->left = 0;
-		if (f->left + gfx->hdisplay > img->i.width)
-		    f->left = img->i.width - gfx->hdisplay;
-	    }
-	    if (blend_msecs && prev && prev != f &&
-		flist_img_get(prev) && flist_img_get(f)) {
-		effect_blend(prev, f);
-		prev = NULL;
-	    } else {
-		shadow_draw_image(img, f->left, f->top, 0, gfx->vdisplay-1, 100);
-	    }
-	    status_update(desc, info);
-	    shadow_render(gfx);
-
-	    if (read_ahead) {
-		struct flist *f = flist_next(fcurrent,1,0);
-		if (f && !f->fimg)
-		    flist_img_load(f,1);
-		status_update(desc, info);
-		shadow_render(gfx);
-	    }
-	}
+        if (redraw) {
+            redraw = 0;
+            if (img->i.height <= gfx->vdisplay) {
+                f->top = 0;
+            } else {
+                if (f->top < 0)
+                    f->top = 0;
+                if (f->top + gfx->vdisplay > img->i.height)
+                    f->top = img->i.height - gfx->vdisplay;
+            }
+            if (img->i.width <= gfx->hdisplay) {
+                f->left = 0;
+            } else {
+                if (f->left < 0)
+                    f->left = 0;
+                if (f->left + gfx->hdisplay > img->i.width)
+                    f->left = img->i.width - gfx->hdisplay;
+            }
+            if (blend_msecs && prev && prev != f &&
+                flist_img_get(prev) && flist_img_get(f)) {
+                effect_blend(prev, f);
+                prev = NULL;
+            } else {
+                shadow_draw_image(img, f->left, f->top, 0, gfx->vdisplay-1, 100);
+            }
+            status_update(desc, info);
+            shadow_render(gfx);
+
+            if (read_ahead) {
+                struct flist *f = flist_next(fcurrent,1,0);
+                if (f && !f->fimg)
+                    flist_img_load(f,1);
+                status_update(desc, info);
+                shadow_render(gfx);
+            }
+        }
         if (check_console_switch()) {
-	    continue;
-	}
-
-	if (!interactive) {
-	    sleep(timeout);
-	    return -1;
-	}
+            continue;
+        }
 
+        if (!interactive) {
+            sleep(timeout);
+            return -1;
+        }
+        // Got Logo Rendered
+        update_post_code(desc, info);
         rc = kbd_wait(paused ? 0 : timeout);
         if (check_console_switch()) {
-	    continue;
-	}
-	if (rc < 1)
-	    return -1; /* timeout */
-    }
-}
-
-static void scale_fix_top_left(struct flist *f, float old, float new)
-{
-    struct ida_image *img = flist_img_get(f);
-    unsigned int width, height;
-    float cx,cy;
-
-    cx = (float)(f->left + gfx->hdisplay/2) / (img->i.width  * old);
-    cy = (float)(f->top  + gfx->vdisplay/2) / (img->i.height * old);
-
-    width   = img->i.width  * new;
-    height  = img->i.height * new;
-    f->left = cx * width  - gfx->hdisplay/2;
-    f->top  = cy * height - gfx->vdisplay/2;
-
-    if (textreading) {
-        f->text_steps = calculate_text_steps(height, gfx->vdisplay);
+            continue;
+        }
+        if (rc < 1)
+            return -1; /* timeout */
     }
 }
 
-/* ---------------------------------------------------------------------- */
-
 static char *my_basename(char *filename)
 {
     char *h;
 
     h = strrchr(filename,'/');
     if (h)
-	return h+1;
+        return h+1;
     return filename;
 }
 
@@ -754,11 +887,11 @@ static char *file_desktop(char *filename)
 
     strncpy(desc,filename,sizeof(desc)-1);
     if (NULL != (h = strrchr(filename,'/'))) {
-	snprintf(desc,sizeof(desc),"%.*s/%s",
-		 (int)(h - filename), filename,
-		 ".directory");
+        snprintf(desc,sizeof(desc),"%.*s/%s",
+                 (int)(h - filename), filename,
+                 ".directory");
     } else {
-	strcpy(desc,".directory");
+        strcpy(desc,".directory");
     }
     return desc;
 }
@@ -774,16 +907,16 @@ static char *make_desc(struct ida_image_info *img, char *filename)
     strncpy(linebuffer,filename,sizeof(linebuffer)-1);
 
     if (comments) {
-	extra = load_find_extra(img, EXTRA_COMMENT);
-	if (extra)
-	    snprintf(linebuffer,sizeof(linebuffer),"%.*s",
-		     extra->size,extra->data);
+        extra = load_find_extra(img, EXTRA_COMMENT);
+        if (extra)
+            snprintf(linebuffer,sizeof(linebuffer),"%.*s",
+                     extra->size,extra->data);
     } else {
-	desc = file_desktop(filename);
-	len = desktop_read_entry(desc, "Comment=", linebuffer, sizeof(linebuffer));
-	if (0 != len)
-	    snprintf(linebuffer+len,sizeof(linebuffer)-len,
-		     " (%s)", my_basename(filename));
+        desc = file_desktop(filename);
+        len = desktop_read_entry(desc, "Comment=", linebuffer, sizeof(linebuffer));
+        if (0 != len)
+            snprintf(linebuffer+len,sizeof(linebuffer)-len,
+                     " (%s)", my_basename(filename));
     }
     return linebuffer;
 }
@@ -793,11 +926,11 @@ static char *make_info(struct ida_image *img, float scale)
     static char linebuffer[128];
 
     snprintf(linebuffer, sizeof(linebuffer),
-	     "%s%.0f%% %ux%u %d/%d",
-	     fcurrent->tag ? "* " : "",
-	     scale*100,
-	     img->i.width, img->i.height,
-	     fcurrent->nr, fcount);
+             "%s%.0f%% %ux%u %d/%d",
+             fcurrent->tag ? "* " : "",
+             scale*100,
+             img->i.width, img->i.height,
+             fcurrent->nr, fcount);
     return linebuffer;
 }
 
@@ -806,19 +939,19 @@ static char *make_info(struct ida_image *img, float scale)
 static struct ida_image *flist_img_get(struct flist *f)
 {
     if (1 != f->scale)
-	return f->simg;
+        return f->simg;
     else
-	return f->fimg;
+        return f->fimg;
 }
 
 static void flist_img_free(struct flist *f)
 {
     if (!f->fimg)
-	return;
+        return;
 
     free_image(f->fimg);
     if (f->simg)
-	free_image(f->simg);
+        free_image(f->simg);
     f->fimg = NULL;
     f->simg = NULL;
     list_del(&f->lru);
@@ -830,7 +963,7 @@ static int flist_img_free_lru(void)
     struct flist *f;
 
     if (img_cnt <= min_cnt)
-	return -1;
+        return -1;
     f = list_entry(flru.next, struct flist, lru);
     flist_img_free(f);
     return 0;
@@ -841,15 +974,15 @@ static void flist_img_release_memory(void)
     int try_release;
 
     for (;;) {
-	try_release = 0;
-	if (img_cnt > max_cnt)
-	    try_release = 1;
-	if (img_mem > max_mem_mb * 1024 * 1024)
-	    try_release = 1;
-	if (!try_release)
-	    break;
-	if (0 != flist_img_free_lru())
-	    break;
+        try_release = 0;
+        if (img_cnt > max_cnt)
+            try_release = 1;
+        if (img_mem > max_mem_mb * 1024 * 1024)
+            try_release = 1;
+        if (!try_release)
+            break;
+        if (0 != flist_img_free_lru())
+            break;
     }
     return;
 }
@@ -859,27 +992,27 @@ static void flist_img_scale(struct flist *f, float scale, int prefetch)
     char linebuffer[128];
 
     if (!f->fimg)
-	return;
+        return;
     if (f->simg && f->scale == scale)
-	return;
+        return;
 
     if (f->simg) {
-	free_image(f->simg);
-	f->simg = NULL;
+        free_image(f->simg);
+        f->simg = NULL;
     }
     if (scale != 1) {
-	if (!prefetch) {
-	    snprintf(linebuffer, sizeof(linebuffer),
-		     "scaling (%.0f%%) %s ...",
-		     scale*100, f->name);
-	    status_update(linebuffer, NULL);
-	}
-	f->simg = scale_image(f->fimg,scale);
-	if (!f->simg) {
-	    snprintf(linebuffer,sizeof(linebuffer),
-		     "%s: scaling FAILED",f->name);
-	    status_error(linebuffer);
-	}
+        if (!prefetch) {
+            snprintf(linebuffer, sizeof(linebuffer),
+                     "scaling (%.0f%%) %s ...",
+                     scale*100, f->name);
+            status_update(linebuffer, NULL);
+        }
+        f->simg = scale_image(f->fimg,scale);
+        if (!f->simg) {
+            snprintf(linebuffer,sizeof(linebuffer),
+                     "%s: scaling FAILED",f->name);
+            status_error(linebuffer);
+        }
     }
     f->scale = scale;
 }
@@ -890,48 +1023,48 @@ static void flist_img_load(struct flist *f, int prefetch)
     float scale = 1;
 
     if (f->fimg) {
-	/* touch */
-	list_del(&f->lru);
-	list_add_tail(&f->lru, &flru);
-	return;
+        /* touch */
+        list_del(&f->lru);
+        list_add_tail(&f->lru, &flru);
+        return;
     }
 
     snprintf(linebuffer,sizeof(linebuffer),"%s %s ...",
-	     prefetch ? "prefetch" : "loading", f->name);
+             prefetch ? "prefetch" : "loading", f->name);
     status_update(linebuffer, NULL);
     f->fimg = read_image(f->name);
     if (!f->fimg) {
-	snprintf(linebuffer,sizeof(linebuffer),
-		 "%s: loading FAILED",f->name);
-	status_error(linebuffer);
-	return;
+        snprintf(linebuffer,sizeof(linebuffer),
+                 "%s: loading FAILED",f->name);
+        status_error(linebuffer);
+        return;
     }
 
     if (!f->seen) {
-	scale = 1;
-	if (autoup || autodown) {
-	    scale = auto_scale(f->fimg);
-	    if (scale < 1 && !autodown)
-		scale = 1;
-	    if (scale > 1 && !autoup)
-		scale = 1;
-	}
+        scale = 1;
+        if (autoup || autodown) {
+            scale = auto_scale(f->fimg);
+            if (scale < 1 && !autodown)
+                scale = 1;
+            if (scale > 1 && !autoup)
+                scale = 1;
+        }
     } else {
-	scale = f->scale;
+        scale = f->scale;
     }
     flist_img_scale(f, scale, prefetch);
 
     if (!f->seen) {
- 	struct ida_image *img = flist_img_get(f);
-	if (img->i.width > gfx->hdisplay)
-	    f->left = (img->i.width - gfx->hdisplay) / 2;
-	if (img->i.height > gfx->vdisplay) {
-	    f->top = (img->i.height - gfx->vdisplay) / 2;
-	    if (textreading) {
+        struct ida_image *img = flist_img_get(f);
+        if (img->i.width > gfx->hdisplay)
+            f->left = (img->i.width - gfx->hdisplay) / 2;
+        if (img->i.height > gfx->vdisplay) {
+            f->top = (img->i.height - gfx->vdisplay) / 2;
+            if (textreading) {
                 f->text_steps = calculate_text_steps(img->i.height, gfx->vdisplay);
-		f->top = 0;
-	    }
-	}
+                f->top = 0;
+            }
+        }
     }
 
     list_add_tail(&f->lru, &flru);
@@ -968,7 +1101,7 @@ static void exit_signals_init(void)
     sigaction(SIGSEGV,&act,&old);
 
     if (0 == (termsig = sigsetjmp(fb_fatal_cleanup,0)))
-	return;
+        return;
 
     /* cleanup */
     gfx->cleanup_display();
@@ -1008,7 +1141,6 @@ int main(int argc, char *argv[])
     bool             framebuffer = false;
     bool             use_libinput;
     char             *info, *desc, *device, *output, *mode;
-    char             linebuffer[128];
     struct flist     *fprev = NULL;
 
 #if 0
@@ -1023,24 +1155,24 @@ int main(int argc, char *argv[])
     cfg_parse_cmdline(&argc,argv,fbi_cfg);
 
     if (GET_AUTO_ZOOM()) {
-	cfg_set_bool(O_AUTO_UP,   1);
-	cfg_set_bool(O_AUTO_DOWN, 1);
+        cfg_set_bool(O_AUTO_UP,   1);
+        cfg_set_bool(O_AUTO_DOWN, 1);
     }
 
     if (GET_HELP()) {
-	usage(stdout, argv[0]);
-	exit(0);
+        usage(stdout, argv[0]);
+        exit(0);
     }
     if (GET_VERSION()) {
-	version();
-	exit(0);
+        version();
+        exit(0);
     }
     if (GET_DEVICE_INFO()) {
         drm_info(cfg_get_str(O_DEVICE));
         exit(0);
     }
     if (GET_WRITECONF())
-	fbi_write_config();
+        fbi_write_config();
 
     once        = GET_ONCE();
     autoup      = GET_AUTO_UP();
@@ -1067,22 +1199,22 @@ int main(int argc, char *argv[])
     filelist    = cfg_get_str(O_FILE_LIST);
 
     if (filelist)
-	flist_add_list(filelist);
+        flist_add_list(filelist);
     for (i = optind; i < argc; i++)
-	flist_add(argv[i]);
+        flist_add(argv[i]);
     flist_renumber();
 
     if (0 == fcount) {
-	usage(stderr, argv[0]);
-	exit(1);
+        usage(stderr, argv[0]);
+        exit(1);
     }
 
     if (GET_RANDOM())
-	flist_randomize();
+        flist_randomize();
     fcurrent = flist_first();
 
     if (NULL == fontname)
-	fontname = "monospace:size=16";
+        fontname = "monospace:size=16";
 
     /* gfx device init */
     device = cfg_get_str(O_DEVICE);
@@ -1135,30 +1267,30 @@ int main(int argc, char *argv[])
     desc = NULL;
     info = NULL;
     for (;;) {
-	flist_img_load(fcurrent, 0);
-	flist_img_release_memory();
-	img = flist_img_get(fcurrent);
-	if (img) {
-	    //desc = make_desc(&fcurrent->fimg->i, fcurrent->name);
-	    //info = make_info(fcurrent->fimg, fcurrent->scale);
-	}
-
-	key = svga_show(fcurrent, fprev, timeout, desc, info, &arg);
-	fprev = fcurrent;
-
-	switch (key) {
-	case -1: /* timeout */
+        flist_img_load(fcurrent, 0);
+        flist_img_release_memory();
+        img = flist_img_get(fcurrent);
+        if (img) {
+            desc = make_desc(&fcurrent->fimg->i, fcurrent->name);
+            info = make_info(fcurrent->fimg, fcurrent->scale);
+        }
+
+        key = svga_show(fcurrent, fprev, timeout, desc, info, &arg);
+        fprev = fcurrent;
+
+        switch (key) {
+        case -1: /* timeout */
             if (filelist && !once && flist_islast(fcurrent)) {
                 flist_check_reload_list(filelist);
                 fcurrent = flist_first();
             } else {
                 fcurrent = flist_next(fcurrent,once,1);
             }
-	    if (NULL == fcurrent) {
-		cleanup_and_exit(0);
-	    }
-	    /* FIXME: wrap around */
-	    break;
-	}
+            if (NULL == fcurrent) {
+                cleanup_and_exit(0);
+            }
+            /* FIXME: wrap around */
+            break;
+        }
     }
 }
-- 
2.25.1

